plugins/module_utils/ndi.py

```diff
@@ -0,0 +1,464 @@
+        elif prefix == "destgrp":
+            return "spanDestGrp"
+        elif prefix == "srcgrp":
+            return "spanSrcGrp"
+        elif prefix == "spanlbl":
+            return "spanSpanLbl"
+        elif prefix == "ctx":
+            return "fvCtx"
+        else:
+            return False
+
+    def construct_tree(self, item_list):
+        """
+        Given a flat list of items, each with a dn. Construct a tree representing their relative relationships.
+        E.g. Given [/a/b/c/d, /a/b, /a/b/c/e, /a/f, /z], the function will construct
+        __root__
+          - a (no data)
+             - b (data of /a/b)
+               - c (no data)
+                 - d (data of /a/b/c/d)
+                 - e (data of /a/b/c/e)
+             - f (data of /a/f)
+          - z (data of /z)
+        __root__ is a predefined name, you could replace this with a flag root:True/False
+        """
+        tree = {'data': None, 'name': '__root__', 'children': {}}
+
+        for item in item_list:
+            for nm, desc in item.items():
+                if 'attributes' not in desc:
+                    raise AssertionError("attributes not in desc")
+                attr = desc.get('attributes')
+                if 'dn' not in attr:
+                    raise AssertionError("dn not in desc")
+                if 'children' in desc:
+                    existing_children = desc.get('children')
+                    self.cmap[attr['dn']] = existing_children
+                path = self.parse_path(attr['dn'])
+                cursor = tree
+                curr_node_dn = ""
+                for node in path:
+                    curr_node_dn += "/" + str(node)
+                    if curr_node_dn[0] == "/":
+                        curr_node_dn = curr_node_dn[1:]
+                    if node not in cursor['children']:
+                        if node == 'uni':
+                            cursor['children'][node] = {
+                                'data': None,
+                                'name': node,
+                                'children': {}
+                            }
+                        else:
+                            aci_class_identifier = node.split("-")[0]
+                            aci_class = self.get_aci_class(
+                                aci_class_identifier)
+                            if not aci_class:
+                                return False
+                            data_dic = {}
+                            data_dic['attributes'] = dict(dn=curr_node_dn)
+                            cursor['children'][node] = {
+                                'data': (aci_class, data_dic),
+                                'name': node,
+                                'children': {}
+                            }
+                    cursor = cursor['children'][node]
+                cursor['data'] = (nm, desc)
+                cursor['name'] = path[-1]
+
+        return tree
+
+    def parse_path(self, dn):
```

Comment:
I think your function will lead to empty string being added to the beginning of the list. Thus your list will be ['', 'a[b/c/d]', 'b', 'c', 'd', 'e'] instead of desired (if I follow your comments) ['a[b/c/d]', 'b', 'c', 'd', 'e'].

Also is there a possibility of data structure that include nested [], a[b/c/d][e/d/f] or a[b/c[d/e/d]f]?

Reply:
Will try to improve when refactor codebase